aes-gcm/src/xaes.rs

```diff
@@ -0,0 +1,127 @@
+use core::ops::{Div, Mul};
+
+use aead::{array::Array, AeadCore, AeadInPlace, Error, Key, KeyInit, KeySizeUser};
+use cipher::{consts::U2, BlockCipherEncrypt, BlockSizeUser};
+
+use crate::{Aes256, Aes256Gcm, Nonce, Tag};
+
+/// XAES-256-GCM
+#[derive(Clone)]
+pub struct XaesGcm256 {
+    aes: Aes256,
+    k1: Block,
+}
+
+type KeySize = <Aes256Gcm as KeySizeUser>::KeySize;
+type NonceSize = <<Aes256Gcm as AeadCore>::NonceSize as Mul<U2>>::Output;
+type TagSize = <Aes256Gcm as AeadCore>::TagSize;
+type CiphertextOverhead = <Aes256Gcm as AeadCore>::CiphertextOverhead;
+type Block = Array<u8, <Aes256 as BlockSizeUser>::BlockSize>;
+
+/// Maximum length of plaintext.
+pub const P_MAX: u64 = 1 << 36;
+
+/// Maximum length of associated data.
+// pub const A_MAX: u64 = 1 << 61;
+pub const A_MAX: u64 = 1 << 36;
+
+/// Maximum length of ciphertext.
+pub const C_MAX: u64 = (1 << 36) + 16;
+
+impl AeadCore for XaesGcm256 {
+    type NonceSize = NonceSize;
+    type TagSize = TagSize;
+    type CiphertextOverhead = CiphertextOverhead;
+}
+
+impl KeySizeUser for XaesGcm256 {
+    type KeySize = KeySize;
+}
+
+impl KeyInit for XaesGcm256 {
+    // Implements step 1 and 2 of the spec.
+    fn new(key: &Key<Aes256>) -> Self {
+        let aes = Aes256::new(key);
+
+        // L = AES-256ₖ(0¹²⁸)
+        let mut k1 = Block::default();
+        aes.encrypt_block(&mut k1);
+
+        // If MSB₁(L) = 0 then K1 = L << 1 Else K1 = (L << 1) ⊕ 0¹²⁰10000111
+        let mut msb = 0;
+        for i in (0..k1.len()).rev() {
```

Comment:
Also probably good to document `k1.len()` conditional here

Reply:
Same as above.